src/pathfinding_service/api/rest.py

```diff
@@ -120,6 +147,102 @@ def post(self, token_network_address: str):
         return {'result': paths}, 200
 
 
+def process_payment(iou_dict, pathfinding_service):
+    if pathfinding_service.service_fee == 0:
+        return
+    if iou_dict is None:
+        raise exceptions.MissingIOU
+
+    # Basic IOU validity checks
+    iou, errors = IOU.Schema().load(iou_dict)
+    if errors:
+        raise exceptions.InvalidRequest(**errors)
+    if iou.receiver != pathfinding_service.address:
+        raise exceptions.WrongIOURecipient(expected=pathfinding_service.address)
+    if not iou.is_signature_valid():
+        raise exceptions.InvalidSignature
+
+    # Compare with known IOU
+    active_iou = pathfinding_service.database.get_iou(
+        sender=iou.sender,
+        claimed=False,
+    )
+    if active_iou:
+        if active_iou.expiration_block != iou.expiration_block:
+            raise exceptions.UseThisIOU(iou=active_iou)
+
+        expected_amount = active_iou.amount + pathfinding_service.service_fee
+    else:
+        if pathfinding_service.database.get_iou(
+            sender=iou.sender,
+            expiration_block=iou.expiration_block,
+            claimed=True,
+        ):
+            raise exceptions.IOUAlreadyClaimed
+
+        min_expiry = pathfinding_service.web3.eth.blockNumber + MIN_IOU_EXPIRY
+        if iou.expiration_block < min_expiry:
+            raise exceptions.IOUExpiredTooEarly(min_expiry=min_expiry)
+        expected_amount = pathfinding_service.service_fee
+    if iou.amount < expected_amount:
+        raise exceptions.InsufficientServicePayment(expected_amount=expected_amount)
+
+    # TODO: deposit large enough?
+
+    # Save latest IOU
+    iou.claimed = False
+    pathfinding_service.database.upsert_iou(iou)
+
+
+@add_schema
+@dataclass
+class IOURequest:
+    """A HTTP request to IOUResource"""
+    sender: Address
+    receiver: Address
+    timestamp: datetime
+    timestamp_str: str = field(metadata={
+        "marshmallow_field": marshmallow.fields.String(load_from='timestamp'),
+    })
+    signature: Signature = field(metadata={"marshmallow_field": HexedBytes()})
+    Schema: ClassVar[Type[marshmallow.Schema]]
+
+    def is_signature_valid(self):
+        packed_data = (
+            Web3.toBytes(hexstr=self.sender) +
+            Web3.toBytes(hexstr=self.receiver) +
+            Web3.toBytes(text=self.timestamp_str)
+        )
+        try:
+            recovered_address = eth_recover(packed_data, self.signature)
+        except InvalidSignature:
+            return False
+        return to_checksum_address(recovered_address) == to_checksum_address(self.sender)
```

Comment:
There's also `eth_utils.is_same_address` which could make this clearer.

Reply:
Good to know. That's better, of course.